### PR TITLE
Upgrade APIGW Exports Nullability Exception Integration

### PR DESCRIPTION
**Pull request by syall**
For changes to files under the `/codegen/aws-models` folder, and manual edits to autogenerated code (e.g. `/service/s3/api.go`) please create an Issue instead of a PR for those type of changes.

N/A.

---

If the PR addresses an existing bug or feature, please reference it here.

Upgrades the integration added in https://github.com/aws/aws-sdk-go-v2/pull/2056.

---

To help speed up the process and reduce the time to merge please ensure that `Allow edits by maintainers` is checked before submitting your PR. This will allow the project maintainers to make minor adjustments or improvements to the submitted PR, allow us to reduce the roundtrip time for merging your request.

N/A.

---

## Description

Due to internal model fixes for API Gateway (APIGW) exports, certain shapes
and members that target those shapes that used to have defaults became
nullable. In order to not break the existing Go v2 client API nullability,
this customization adds back default values for affected shapes.

Nullability exceptions are root shapes captured in `APIGW_exports_nullability_exceptions.json`.

Definitions:

- “defaulted” shapes are shapes that have the default trait applied.
- “non-defaulted” shapes are shapes that have the default trait applied.
- “Existing” shapes are shapes that are present in a previous version of a model
- “Added” shapes are shapes that are new in a current version of a model, not present in a previous version of a model

Shape scenarios we need to be aware of when defining the customization:

- Existing “defaulted” root boolean or number shapes MUST be backfilled with a default trait.
- TODO(APIGW): remove after model fixes
  - Added root boolean or number shapes with default traits MUST be backfilled with a default trait.
    - This is due to not having context of the previous models: there is no way to tell if the added root shape will be affected by the C2J model fixes
- TODO(APIGW): will be true after model fixes
  - Added root boolean or number shapes with default traits MUST NOT be backfilled with a default trait.
- Added root boolean or number shapes without default traits MUST NOT be backfilled with a default trait.
- Existing member shapes that target a “defaulted” root shape MUST be backfilled with a default trait.
- Existing member shapes that target a “non-defaulted” root shape MUST NOT be backfilled with a default trait.
- Existing member shapes that change targets from a “defaulted” root shape to a “non-defaulted” root shape will throw an error.
  - TODO(APIGW): provide a green path forward?
- Existing member shapes that change targets from a “non-defaulted” root shape to a “defaulted” root shape will throw an error.
  - TODO(APIGW): provide a green path forward?
- Added member shapes that target a “defaulted” root shape MUST be backfilled with a default trait.
- Added member shapes that target a “non-defaulted” root shape MUST NOT be backfilled with a default trait.

## Testing

1. `make generate`
2. `cd codegen/ && ./gradlew :smithy-aws-go-codegen:check`
3. Tested with the fixed version of `medialive.json` locally

### CODE REVIEW

**Review by isaiahvita — APPROVED** [review]

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java [hunk 1]
```diff
@@ -110,150 +138,269 @@ public Model preprocessModel(Model model, GoSettings settings) {
...
-                nullabilityExceptions, memberShapes);
+        String sdkId = model.getServiceShapesWithTrait(ServiceTrait.class).iterator().next()
+                .expectTrait(ServiceTrait.class).getSdkId()
+                .replace(" ", "-")
+                .toLowerCase();
 
-        // Replace nullability exception shapes
-        return ModelTransformer.create().replaceShapes(model, shapesToReplace);
+        String sha = getPreviousModelSha(root, sdkId, service);
+        LOGGER.warning("Previous model SHA for " + service.toString() + ": " + sha);
```
Comment (hpmellema) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java:185]: If these log statements going to be the generator for a while I would recommend downgrading these to DEBUG so they don't pollute logs by default. It makes warnings less useful when debug data like these are mixed in.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AwsSdkGoV2ChangedNullability.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        ShapeId service = newModel
+                .getServiceShapesWithTrait(ServiceTrait.class)
+                .iterator().next()
+                .toShapeId();
+        Set<ShapeId> nullabilityExceptions = getNullabilityExceptions(service);
+
+        differences.addedShapes().forEach(shape -> {
+            // TODO(APIGW): remove after model fixes
+            if (shape.hasTrait(DefaultTrait.class)
+                    && (shape.isBooleanShape() || newNumberShapes.contains(shape))
```
Comment (hpmellema) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AwsSdkGoV2ChangedNullability.java:67]: I dont think this actually catches all of the cases we want. It doesnt check if there is a member shape marked with a default value that targets a number or boolean shape.

I think using selectors here might be more robust. I think the following two selectors cover all cases: 
```
member[trait|smithy.api#default]:test(> :is(boolean, number))
```
```
[trait|smithy.api#default]:is(boolean, number)
```

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AwsSdkGoV2ChangedNullability.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                    && !nullabilityExceptions.contains(shape.toShapeId())) {
+                events.add(error(shape, "Shape must be added to the APIGW exports nullability exceptions"));
+            }
+        });
+
+        differences.changedShapes().forEach(changedShape -> {
+            ShapeId shape = changedShape.getShapeId();
```
Comment (hpmellema) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AwsSdkGoV2ChangedNullability.java:74]: thought: 
I personally like doing this as: 
```
differences.changedShapes().map(Shape::getShapeId).forEach()
```

File: .github/workflows/codegen.yml [hunk 4]
```diff
@@ -18,6 +18,8 @@ jobs:
       JAVA_TOOL_OPTIONS: "-Xmx2g"
     steps:
     - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (hpmellema) [reply .github/workflows/codegen.yml:22]: This still seems wrong but is an issue with the github action. Seems like this action cannot pull all for a single branch.

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java [hunk 5]
```diff
@@ -15,69 +15,97 @@
...
+ *   shape to a “defaulted” root shape will throw an error.
+ *   - TODO(APIGW): provide a green path forward?
+ * - Added member shapes that target a “defaulted” root shape MUST be backfilled
+ *   with a default trait.
+ * - Added member shapes that target a “non-defaulted” root shape MUST NOT be
+ *   backfilled with a default trait.
  */
 @SmithyInternalApi
 public class ApiGatewayExportsNullabilityExceptionIntegration implements GoIntegration {
+    static final String NULLABILITY_EXCEPTIONS_FILE = "APIGW_exports_nullability_exceptions.json";
```
[the remark this answers is not in the document]
Comment (syall) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java:106]: This is exposed to package-private for the test class.

File: .github/workflows/codegen.yml
[the remark this answers is not in the document]
Comment (syall) [reply .github/workflows/codegen.yml:1]: Created an internal story to track this.

**Review by hpmellema — APPROVED** [review]

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java (hunk 1 above)
Comment (syall) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/ApiGatewayExportsNullabilityExceptionIntegration.java]: Changed to `LOGGER.info()`.

**Pull request merged**